tests: pandas and dask backend tests are failing

These pandas (and derived) backend tests are currently failing on `master` due to the `IBIS_TEST_DATA_DIRECTORY` environment variable not being set:

These are failures to fix (or run) tests, not broken parts of the library:

```
FAILED ibis/backends/pandas/tests/execution/test_operations.py::test_summary_numeric - AttributeError: 'list' object has no attribute 'execute'
FAILED ibis/backends/pandas/tests/execution/test_window.py::test_batting_most_hits - ibis.common.exceptions.OperationNotDefinedError: Window functions affected by following with order_by are not implemented
FAILED ibis/backends/pandas/tests/execution/test_operations.py::test_summary_non_numeric - AttributeError: 'list' object has no attribute 'execute'
```

These tests never worked because the data never existed for them:

```
ERROR ibis/backends/tests/test_join.py::test_join_project_left_table[parquet-anti] - AttributeError: batting
ERROR ibis/backends/tests/test_join.py::test_join_project_left_table[parquet-outer] - AttributeError: batting
ERROR ibis/backends/tests/test_join.py::test_join_project_left_table[parquet-right] - AttributeError: batting
ERROR ibis/backends/tests/test_join.py::test_join_project_left_table[parquet-semi] - AttributeError: batting
ERROR ibis/backends/tests/test_join.py::test_join_project_left_table[parquet-left] - AttributeError: batting
ERROR ibis/backends/tests/test_join.py::test_join_project_left_table[parquet-inner] - AttributeError: batting
```

These tests never worked, so they don't block the release:

```
ERROR ibis/backends/dask/tests/execution/test_operations.py::test_where_long - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
ERROR ibis/backends/dask/tests/execution/test_operations.py::test_searched_case_column - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
ERROR ibis/backends/dask/tests/execution/test_operations.py::test_summary_numeric_group_by - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
ERROR ibis/backends/dask/tests/execution/test_operations.py::test_summary_non_numeric - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
ERROR ibis/backends/dask/tests/execution/test_operations.py::test_quantile_groupby - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
ERROR ibis/backends/dask/tests/execution/test_join.py::test_join_with_window_function - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
ERROR ibis/backends/dask/tests/execution/test_operations.py::test_summary_non_numeric_group_by - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
ERROR ibis/backends/dask/tests/execution/test_operations.py::test_simple_case_column - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
ERROR ibis/backends/dask/tests/execution/test_operations.py::test_summary_numeric - NotImplementedError: 'DataFrame.iloc' only supports selecting columns. It must be used like 'df.iloc[:, column_indexer]'.
```

--- a/ibis/backends/dask/tests/execution/conftest.py
+++ b/ibis/backends/dask/tests/execution/conftest.py
@@ -1,5 +1,4 @@
 import decimal
-import os
 
 import dask.dataframe as dd
 import pandas as pd
@@ -72,31 +71,21 @@ def df(npartitions):
 
 
 @pytest.fixture(scope='module')
-def batting_df():
-    path = os.path.join(
-        os.environ.get('IBIS_TEST_DATA_DIRECTORY', ''), 'batting.csv'
+def batting_df(data_directory):
+    df = dd.read_csv(
+        data_directory / 'batting.csv',
+        assume_missing=True,
+        dtype={"lgID": "object"},
     )
-    if not os.path.exists(path):
-        pytest.skip(f'{path} not found')
-    elif not os.path.isfile(path):
-        pytest.skip(f'{path} is not a file')
-
-    df = dd.read_csv(path, assume_missing=True)
-    num_rows = int(0.01 * len(df))
-    return df.iloc[30 : 30 + num_rows].reset_index(drop=True)
+    return df.sample(frac=0.01).reset_index(drop=True)
 
 
 @pytest.fixture(scope='module')
-def awards_players_df():
-    path = os.path.join(
-        os.environ.get('IBIS_TEST_DATA_DIRECTORY', ''), 'awards_players.csv'
+def awards_players_df(data_directory):
+    return dd.read_csv(
+        data_directory / 'awards_players.csv',
+        assume_missing=True,
     )
-    if not os.path.exists(path):
-        pytest.skip(f'{path} not found')
-    elif not os.path.isfile(path):
-        pytest.skip(f'{path} is not a file')
-
-    return dd.read_csv(path, assume_missing=True)
 
 
 @pytest.fixture(scope='module')

--- a/ibis/backends/dask/tests/execution/test_join.py
+++ b/ibis/backends/dask/tests/execution/test_join.py
@@ -305,6 +305,10 @@ def test_join_with_project_right_duplicate_column(client, how, left, df1, df3):
     )
 
 
+@pytest.mark.xfail(
+    raises=NotImplementedError,
+    reason="multi-key sort isn't implemented",
+)
 def test_join_with_window_function(
     players_base, players_df, batting, batting_df
 ):

--- a/ibis/backends/dask/tests/execution/test_operations.py
+++ b/ibis/backends/dask/tests/execution/test_operations.py
@@ -790,6 +790,10 @@ def test_round(t, df):
     npt.assert_almost_equal(result, expected, decimal=precision)
 
 
+@pytest.mark.xfail(
+    raises=NotImplementedError,
+    reason="MultiQuantile is not implemented for the dask backend",
+)
 def test_quantile_groupby(batting, batting_df):
     def q_fun(x, quantile, interpolation):
         res = x.quantile(quantile, interpolation=interpolation).tolist()
@@ -811,8 +815,8 @@ def test_quantile_groupby(batting, batting_df):
 
 
 def test_summary_numeric(batting, batting_df):
-    expr = batting.G.summary()
-    result = expr.compile()
+    expr = batting.aggregate(batting.G.summary())
+    result = expr.execute()
     assert len(result) == 1
 
     G = batting_df.G
@@ -830,26 +834,24 @@ def test_summary_numeric(batting, batting_df):
 
 def test_summary_numeric_group_by(batting, batting_df):
     expr = batting.groupby('teamID').G.summary()
-    result = expr.compile()
+    result = expr.execute()
     expected = (
         batting_df.groupby('teamID')
         .G.apply(
-            lambda s: dd.from_pandas(
-                pd.DataFrame(
-                    {
-                        'count': s.count(),
-                        'nulls': s.isnull().sum(),
-                        'min': s.min(),
-                        'max': s.max(),
-                        'sum': s.sum(),
-                        'mean': s.mean(),
-                        'approx_nunique': s.nunique(),
-                    },
-                    index=[0],
-                ),
-                npartitions=1,
+            lambda s: pd.DataFrame(
+                {
+                    'count': s.count(),
+                    'nulls': s.isnull().sum(),
+                    'min': s.min(),
+                    'max': s.max(),
+                    'sum': s.sum(),
+                    'mean': s.mean(),
+                    'approx_nunique': s.nunique(),
+                },
+                index=[0],
             )
         )
+        .compute()
         .reset_index(level=1, drop=True)
         .reset_index()
     )
@@ -859,8 +861,8 @@ def test_summary_numeric_group_by(batting, batting_df):
 
 
 def test_summary_non_numeric(batting, batting_df):
-    expr = batting.teamID.summary()
-    result = expr.compile()
+    expr = batting.aggregate(batting.teamID.summary())
+    result = expr.execute()
     assert len(result) == 1
     assert len(result.columns) == 3
     expected = {
@@ -873,22 +875,20 @@ def test_summary_non_numeric(batting, batting_df):
 
 def test_summary_non_numeric_group_by(batting, batting_df):
     expr = batting.groupby('teamID').playerID.summary()
-    result = expr.compile()
+    result = expr.execute()
     expected = (
         batting_df.groupby('teamID')
         .playerID.apply(
-            lambda s: dd.from_pandas(
-                pd.DataFrame(
-                    {
-                        'count': s.count(),
-                        'nulls': s.isnull().sum(),
-                        'uniques': s.nunique(),
-                    },
-                    index=[0],
-                ),
-                npartitions=1,
+            lambda s: pd.DataFrame(
+                {
+                    'count': s.count(),
+                    'nulls': s.isnull().sum(),
+                    'uniques': s.nunique(),
+                },
+                index=[0],
             )
         )
+        .compute()
         .reset_index(level=1, drop=True)
         .reset_index()
     )
@@ -994,7 +994,7 @@ def test_difference(client, df1, intersect_df2):
     merged = df1.merge(
         intersect_df2, on=list(df1.columns), how="outer", indicator=True
     )
-    expected = merged[merged["_merge"] != "both"].drop("_merge", 1)
+    expected = merged[merged["_merge"] != "both"].drop("_merge", axis=1)
 
     # force same index
     result = result.compute().reset_index(drop=True)

--- a/ibis/backends/hdf5/__init__.py
+++ b/ibis/backends/hdf5/__init__.py
@@ -16,6 +16,22 @@ class Backend(BaseFileBackend):
     extension = 'h5'
     table_class = HDFTable
 
+    def do_connect(self, path):
+        """Create an HDF5 backend for use with Ibis
+
+        Parameters
+        ----------
+        path
+            Path to an HDF5 file
+
+        Returns
+        -------
+        Backend
+        """
+        from ibis.backends.pandas import execution, udf  # noqa: F401
+
+        super().do_connect(path)
+
     def list_tables(self, path=None, like=None, database=None):
         """
         For HDF5, tables are the HDF5 tables inside the file.

--- a/ibis/backends/pandas/execution/generic.py
+++ b/ibis/backends/pandas/execution/generic.py
@@ -808,7 +808,7 @@ def execute_difference_dataframe_dataframe(
     merged = left.merge(
         right, on=list(left.columns), how='outer', indicator=True
     )
-    result = merged[merged["_merge"] != "both"].drop("_merge", 1)
+    result = merged[merged["_merge"] != "both"].drop("_merge", axis=1)
     return result
 
 

--- a/ibis/backends/pandas/execution/selection.py
+++ b/ibis/backends/pandas/execution/selection.py
@@ -333,7 +333,7 @@ def build_df_from_selection(
             renamed_cols[from_col] = to_cols[0]
         else:
             for new_col in to_cols:
-                result[new_col] = result[from_col]
+                result[new_col] = result.loc[:, from_col]
 
     if renamed_cols:
         result = result.rename(columns=renamed_cols)

--- a/ibis/backends/pandas/tests/conftest.py
+++ b/ibis/backends/pandas/tests/conftest.py
@@ -4,6 +4,7 @@ import pandas as pd
 
 import ibis
 import ibis.expr.operations as ops
+import ibis.expr.types as ir
 from ibis.backends.tests.base import BackendTest, RoundHalfToEven
 
 
@@ -33,3 +34,15 @@ class TestConf(BackendTest, RoundHalfToEven):
                 ),
             }
         )
+
+    @property
+    def functional_alltypes(self) -> ir.TableExpr:
+        return self.connection.table("functional_alltypes")
+
+    @property
+    def batting(self) -> ir.TableExpr:
+        return self.connection.table("batting")
+
+    @property
+    def awards_players(self) -> ir.TableExpr:
+        return self.connection.table("awards_players")

--- a/ibis/backends/pandas/tests/execution/conftest.py
+++ b/ibis/backends/pandas/tests/execution/conftest.py
@@ -1,5 +1,4 @@
 import decimal
-import os
 
 import numpy as np
 import pandas as pd
@@ -80,31 +79,27 @@ def df():
 
 
 @pytest.fixture(scope='module')
-def batting_df():
-    path = os.path.join(
-        os.environ.get('IBIS_TEST_DATA_DIRECTORY', ''), 'batting.csv'
+def batting_df(data_directory):
+    num_rows = 1000
+    start_index = 30
+    df = pd.read_csv(
+        data_directory / 'batting.csv',
+        index_col=None,
+        sep=',',
+        header=0,
+        skiprows=range(1, start_index + 1),
+        nrows=num_rows,
     )
-    if not os.path.exists(path):
-        pytest.skip(f'{path} not found')
-    elif not os.path.isfile(path):
-        pytest.skip(f'{path} is not a file')
-
-    df = pd.read_csv(path, index_col=None, sep=',')
-    num_rows = int(0.01 * len(df))
-    return df.iloc[30 : 30 + num_rows].reset_index(drop=True)
+    return df.reset_index(drop=True)
 
 
 @pytest.fixture(scope='module')
-def awards_players_df():
-    path = os.path.join(
-        os.environ.get('IBIS_TEST_DATA_DIRECTORY', ''), 'awards_players.csv'
+def awards_players_df(data_directory):
+    return pd.read_csv(
+        data_directory / 'awards_players.csv',
+        index_col=None,
+        sep=',',
     )
-    if not os.path.exists(path):
-        pytest.skip(f'{path} not found')
-    elif not os.path.isfile(path):
-        pytest.skip(f'{path} is not a file')
-
-    return pd.read_csv(path, index_col=None, sep=',')
 
 
 @pytest.fixture(scope='module')

--- a/ibis/backends/pandas/tests/execution/test_operations.py
+++ b/ibis/backends/pandas/tests/execution/test_operations.py
@@ -345,15 +345,10 @@ def test_frame_limit(t, df, offset):
     tm.assert_frame_equal(result[expected.columns], expected)
 
 
-@pytest.mark.xfail(
-    raises=AttributeError, reason='TableColumn does not implement limit'
-)
 @pytest.mark.parametrize('offset', [0, 2])
 def test_series_limit(t, df, offset):
-    n = 5
-    s_expr = t.plain_int64.limit(n, offset=offset)
-    result = s_expr.execute()
-    tm.assert_series_equal(result, df.plain_int64.iloc[offset : offset + n])
+    with pytest.raises(AttributeError):
+        t.plain_int64.limit(5, offset=offset)
 
 
 @pytest.mark.parametrize(
@@ -682,7 +677,7 @@ def test_summary_execute(t):
 
 
 def test_summary_numeric(batting, batting_df):
-    expr = batting.G.summary()
+    expr = batting.aggregate(batting.G.summary())
     result = expr.execute()
     assert len(result) == 1
 
@@ -728,7 +723,7 @@ def test_summary_numeric_group_by(batting, batting_df):
 
 
 def test_summary_non_numeric(batting, batting_df):
-    expr = batting.teamID.summary()
+    expr = batting.aggregate(batting.teamID.summary())
     result = expr.execute()
     assert len(result) == 1
     assert len(result.columns) == 3
@@ -855,7 +850,7 @@ def test_difference(client, df1, intersect_df2):
     merged = df1.merge(
         intersect_df2, on=list(df1.columns), how="outer", indicator=True
     )
-    expected = merged[merged["_merge"] != "both"].drop("_merge", 1)
+    expected = merged[merged["_merge"] != "both"].drop("_merge", axis=1)
     tm.assert_frame_equal(result, expected)
 
 

--- a/ibis/backends/parquet/__init__.py
+++ b/ibis/backends/parquet/__init__.py
@@ -62,6 +62,22 @@ class Backend(BaseFileBackend):
     extension = 'parquet'
     table_class = ParquetTable
 
+    def do_connect(self, path):
+        """Create a Parquet backend for use with Ibis
+
+        Parameters
+        ----------
+        path
+            Path to a parquet file
+
+        Returns
+        -------
+        Backend
+        """
+        from ibis.backends.pandas import execution, udf  # noqa: F401
+
+        super().do_connect(path)
+
     def insert(self, path, expr, **kwargs):
         path = self.root / path
         df = execute(expr)

--- a/ibis/backends/parquet/tests/conftest.py
+++ b/ibis/backends/parquet/tests/conftest.py
@@ -15,9 +15,6 @@ class TestConf(PandasTest):
 
     @staticmethod
     def connect(data_directory: Path):
-        filename = data_directory / 'functional_alltypes.parquet'
-        if not filename.exists():
-            pytest.skip(f'test data set {filename} not found')
         return ibis.parquet.connect(data_directory)
 
 

--- a/ibis/backends/tests/test_join.py
+++ b/ibis/backends/tests/test_join.py
@@ -2,6 +2,8 @@ import pandas as pd
 import pytest
 from pytest import param
 
+import ibis.common.exceptions as exc
+
 # add here backends that passes join tests
 all_db_join_supported = [
     'csv',
@@ -23,13 +25,15 @@ all_db_join_supported = [
         param(
             'semi',
             marks=pytest.mark.xfail(
-                raises=NotImplementedError, reason='Semi join not implemented'
+                raises=(exc.OperationNotDefinedError, NotImplementedError),
+                reason='Semi join not implemented',
             ),
         ),
         param(
             'anti',
             marks=pytest.mark.xfail(
-                raises=NotImplementedError, reason='Anti join not implemented'
+                raises=(exc.OperationNotDefinedError, NotImplementedError),
+                reason='Anti join not implemented',
             ),
         ),
     ],

--- a/ibis/backends/tests/test_join.py
+++ b/ibis/backends/tests/test_join.py
@@ -3,7 +3,14 @@ import pytest
 from pytest import param
 
 # add here backends that passes join tests
-all_db_join_supported = ['pandas', 'pyspark', 'dask', 'csv']
+all_db_join_supported = [
+    'csv',
+    'dask',
+    'hdf5',
+    'pandas',
+    'parquet',
+    'pyspark',
+]
 
 
 @pytest.mark.parametrize(
@@ -28,8 +35,6 @@ all_db_join_supported = ['pandas', 'pyspark', 'dask', 'csv']
     ],
 )
 @pytest.mark.only_on_backends(all_db_join_supported)
-# Csv is a subclass of Pandas so need to skip it explicitly.
-@pytest.mark.xfail_unsupported
 def test_join_project_left_table(backend, con, batting, awards_players, how):
 
     left = batting[batting.yearID == 2015]

--- a/shell.nix
+++ b/shell.nix
@@ -49,11 +49,13 @@ pkgs.mkShell {
   name = "ibis${pythonShortVersion}";
 
   shellHook = ''
-    data_dir="ci/ibis-testing-data"
+    data_dir="$PWD/ci/ibis-testing-data"
     mkdir -p "$data_dir"
     chmod u+rwx "$data_dir"
     cp -rf ${pkgs.ibisTestingData}/* "$data_dir"
     chmod --recursive u+rw "$data_dir"
+
+    export IBIS_TEST_DATA_DIRECTORY="$data_dir"
   '';
 
   buildInputs = devDeps ++ libraryDevDeps ++ [

